File: chinatechnews/CNTechNews.java
package skills.cntechnews;

import org.json.JSONArray;
import org.json.JSONObject;
import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserFactory;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.StringReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * AutoClaw CN Tech News Skill
 * Aggregates top 20 Chinese tech news from 5 sources via RSS.
 * No external dependencies. Android built-in APIs only.
 * Content in Chinese.
 *
 * Sources (RSS):
 *   36kr    = 36氪 (36kr.com/feed)
 *   tmt     = 钛媒体 (tmtpost.com/rss.xml)
 *   ithome  = IT之家 (ithome.com/rss/)
 *   sspai   = 少数派 (sspai.com/feed)
 *   sina    = 新浪科技 (rss.sina.com.cn)
 *
 * Usage:
 *   CNTechNews news = new CNTechNews();
 *   news.fetchAll(new CNTechNews.Callback() { ... });
 *   news.fetchAll(10, new String[]{"36kr","ithome"}, new CNTechNews.Callback() { ... });
 */
public class CNTechNews {

    private static final int TIMEOUT = 15000;
    private static final String UA = "AutoClaw-CNTechNews/1.0";
    private static final int DEFAULT_COUNT = 20;

    private final ExecutorService executor = Executors.newCachedThreadPool();

    // ── Callback ──────────────────────────────────────────────

    public interface Callback {
        void onSuccess(NewsResult result);
        void onError(String error);
    }

    // ── Data Models ───────────────────────────────────────────

    public static class Article {
        public String source;
        public String title;
        public String url;
        public String description;
        public String published;
        public String author;
        public String error;

        public JSONObject toJson() {
            try {
                JSONObject o = new JSONObject();
                if (error != null) {
                    o.put("source", source);
                    o.put("error", error);
                    return o;
                }
                o.put("source", source);
                o.put("title", title);
                o.put("url", url);
                if (description != null) o.put("description", description);
                if (published != null) o.put("published", published);
                if (author != null) o.put("author", author);
                return o;
            } catch (Exception e) {
                return new JSONObject();
            }
        }
    }

    public static class NewsResult {
        public Map<String, List<Article>> results;
        public String briefSummary;
        public String detailedSummary;

        public JSONObject toJson() {
            try {
                JSONObject o = new JSONObject();
                for (Map.Entry<String, List<Article>> entry : results.entrySet()) {
                    JSONArray arr = new JSONArray();
                    for (Article a : entry.getValue()) arr.put(a.toJson());
                    o.put(entry.getKey(), arr);
                }
                return o;
            } catch (Exception e) {
                return new JSONObject();
            }
        }
    }

    // ── Source Config ─────────────────────────────────────────

    private static final Map<String, String> SOURCE_NAMES = new LinkedHashMap<>();
    static {
        SOURCE_NAMES.put("36kr", "36\u6c2a");
        SOURCE_NAMES.put("tmt", "\u949b\u5a92\u4f53");
        SOURCE_NAMES.put("ithome", "IT\u4e4b\u5bb6");
        SOURCE_NAMES.put("sspai", "\u5c11\u6570\u6d3e");
        SOURCE_NAMES.put("sina", "\u65b0\u6d6a\u79d1\u6280");
    }

    private static final Map<String, String> RSS_FEEDS = new LinkedHashMap<>();
    static {
        RSS_FEEDS.put("36kr", "https://36kr.com/feed");
        RSS_FEEDS.put("tmt", "https://www.tmtpost.com/rss.xml");
        RSS_FEEDS.put("ithome", "https://www.ithome.com/rss/");
        RSS_FEEDS.put("sspai", "https://sspai.com/feed");
        RSS_FEEDS.put("sina", "https://rss.sina.com.cn/tech/rollnews.xml");
    }

    // ── HTML Stripping ────────────────────────────────────────

    private static String stripHtml(String s) {
        if (s == null || s.isEmpty()) return "";
        return s.replaceAll("<[^>]*>", "").replaceAll("\\s+", " ").trim();
    }

    private static String clean(String s) {
        if (s == null || s.isEmpty()) return null;
        s = stripHtml(s);
        return s.length() > 500 ? s.substring(0, 500) : s;
    }

    // ── HTTP ──────────────────────────────────────────────────

    private static String httpGet(String urlStr) throws Exception {
        URL url = new URL(urlStr);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        conn.setRequestProperty("User-Agent", UA);
        conn.setRequestProperty("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8");
        conn.setRequestProperty("Accept-Language", "zh-CN,zh;q=0.9,en;q=0.8");
        conn.setConnectTimeout(TIMEOUT);
        conn.setReadTimeout(TIMEOUT);
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream(), "UTF-8"))) {
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) sb.append(line).append("\n");
            return sb.toString();
        } finally {
            conn.disconnect();
        }
    }

    // ── RSS Parser ────────────────────────────────────────────

    private static List<Article> fetchRss(String feedUrl, String sourceName, int n) {
        List<Article> articles = new ArrayList<>();
        try {
            String xml = httpGet(feedUrl);
            XmlPullParserFactory factory = XmlPullParserFactory.newInstance();
            factory.setNamespaceAware(true);
            XmlPullParser parser = factory.newPullParser();
            parser.setInput(new StringReader(xml));

            boolean inItem = false;
            String currentTag = "";
            String title = "", link = "", description = "", pubDate = "", author = "";

            int eventType = parser.getEventType();
            while (eventType != XmlPullParser.END_DOCUMENT) {
                if (eventType == XmlPullParser.START_TAG) {
                    String tag = parser.getName();
                    if ("item".equals(tag) || "entry".equals(tag)) {
                        inItem = true;
                        title = "";
                        link = "";
                        description = "";
                        pubDate = "";
                        author = "";
                    }
                    if (inItem) {
                        currentTag = tag;
                        if ("link".equals(tag)) {
                            String href = parser.getAttributeValue(null, "href");
                            if (href != null && !href.isEmpty()) link = href;
                        }
                    }
                } else if (eventType == XmlPullParser.TEXT && inItem) {
                    String text = parser.getText();
                    if (text != null) text = text.trim();
                    if (text != null && !text.isEmpty()) {
                        switch (currentTag) {
                            case "title":
                                title += text;
                                break;
                            case "link":
                                if (link.isEmpty()) link = text;
                                break;
                            case "description":
                                description += text;
                                break;
                            case "summary":
                                description += text;
                                break;
                            case "pubDate":
                                pubDate += text;
                                break;
                            case "published":
                                pubDate += text;
                                break;
                            case "updated":
                                if (pubDate.isEmpty()) pubDate += text;
                                break;
                            case "creator":
                                author += text;
                                break;
                            case "author":
                                author += text;
                                break;
                            case "encoded":
                                String cleaned = clean(text);
                                if (cleaned != null && cleaned.length() > description.length())
                                    description = cleaned;
                                break;
                        }
                    }
                } else if (eventType == XmlPullParser.END_TAG) {
                    String tag = parser.getName();
                    if (("item".equals(tag) || "entry".equals(tag)) && inItem) {
                        inItem = false;
                        title = title.trim();
                        if (!title.isEmpty() && articles.size() < n) {
                            Article a = new Article();
                            a.source = sourceName;
                            a.title = title;
                            a.url = link.trim();
                            String desc = clean(description);
                            if (desc != null && !desc.isEmpty())
                                a.description = desc.length() > 300 ? desc.substring(0, 300) : desc;
                            if (!pubDate.trim().isEmpty()) a.published = pubDate.trim();
                            if (!author.trim().isEmpty()) a.author = author.trim();
                            articles.add(a);
                        }
                    }
                    if (tag.equals(currentTag)) currentTag = "";
                }
                eventType = parser.next();
            }
        } catch (Exception e) {
            Article err = new Article();
            err.source = sourceName;
            err.error = e.getMessage();
            articles.add(err);
        }
        return articles;
    }

    // ── Formatters ────────────────────────────────────────────

    private static String fmtBrief(Map<String, List<Article>> allResults) {
        StringBuilder sb = new StringBuilder();
        int idx = 1;
        for (Map.Entry<String, List<Article>> entry : allResults.entrySet()) {
            for (Article a : entry.getValue()) {
                if (a.error != null || a.title == null || a.title.isEmpty()) continue;
                sb.append(idx).append(". [").append(entry.getKey()).append("] ").append(a.title);
                if (a.description != null) {
                    String desc = a.description.length() > 150 ? a.description.substring(0, 150) : a.description;
                    sb.append(" -- ").append(desc);
                }
                sb.append("\n");
                idx++;
            }
        }
        String now = new SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.CHINA).format(new Date());
        return "\u4e2d\u56fd\u79d1\u6280\u65b0\u95fb (" + now + ") - " + (idx - 1) + " \u7bc7\u6587\u7ae0\n" + sb;
    }

    private static String fmtDetailed(Map<String, List<Article>> allResults) {
        StringBuilder sb = new StringBuilder();
        int total = 0;
        for (Map.Entry<String, List<Article>> entry : allResults.entrySet()) {
            List<Article> valid = new ArrayList<>();
            List<Article> errors = new ArrayList<>();
            for (Article a : entry.getValue()) {
                if (a.error != null) errors.add(a);
                else if (a.title != null && !a.title.isEmpty()) valid.add(a);
            }
            sb.append("\n## ").append(entry.getKey()).append(" (").append(valid.size()).append(" \u7bc7)\n");
            for (Article e : errors) sb.append("  [\u9519\u8bef: ").append(e.error).append("]\n");
            int i = 1;
            for (Article a : valid) {
                sb.append("  ").append(i).append(". **").append(a.title).append("**\n");
                if (a.url != null && !a.url.isEmpty()) sb.append("     ").append(a.url).append("\n");
                if (a.description != null) {
                    String desc = a.description.length() > 250 ? a.description.substring(0, 250) : a.description;
                    sb.append("     ").append(desc).append("\n");
                }
                if (a.published != null) sb.append("     \u53d1\u5e03: ").append(a.published).append("\n");
                i++;
                total++;
            }
        }
        String now = new SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.CHINA).format(new Date());
        return "# \u4e2d\u56fd\u79d1\u6280\u65b0\u95fb - " + now + "\n**" + total + " \u7bc7\u6587\u7ae0\uff0c\u6765\u81ea " + allResults.size() + " \u4e2a\u6e90**\n" + sb;
    }

    // ── Public API ────────────────────────────────────────────

    /** Fetch all sources, 20 articles each. */
    public void fetchAll(Callback callback) {
        fetchAll(DEFAULT_COUNT, null, callback);
    }

    /** Fetch selected sources with custom count. Pass null for sources to fetch all. */
    public void fetchAll(int count, String[] sources, Callback callback) {
        executor.execute(() -> {
            try {
                Map<String, List<Article>> results = new LinkedHashMap<>();
                String[] targets = sources != null ? sources : SOURCE_NAMES.keySet().toArray(new String[0]);

                for (String key : targets) {
                    String name = SOURCE_NAMES.get(key);
                    String feed = RSS_FEEDS.get(key);
                    if (name != null && feed != null) {
                        results.put(name, fetchRss(feed, name, count));
                    }
                }

                NewsResult result = new NewsResult();
                result.results = results;
                result.briefSummary = fmtBrief(results);
                result.detailedSummary = fmtDetailed(results);
                callback.onSuccess(result);
            } catch (Exception e) {
                callback.onError(e.getMessage());
            }
        });
    }

    /** Fetch a single source by code (36kr, tmt, ithome, sspai, sina). */
    public void fetchSource(String code, int count, Callback callback) {
        fetchAll(count, new String[]{code}, callback);
    }

    /** Shutdown the executor when done. */
    public void shutdown() {
        executor.shutdown();
    }
}

File: chinatechnews/cntechnews.js
/**
 * AutoClaw CN Tech News Skill
 * Aggregates top 20 Chinese tech news from 4 sources via RSS.
 * No external dependencies. Node.js built-in APIs only.
 * Content in Chinese.
 *
 * Sources (RSS):
 *   36kr    = 36氪 (36kr.com/feed)
 *   tmt     = 钛媒体 (tmtpost.com/rss.xml)
 *   ithome  = IT之家 (ithome.com/rss/)
 *   sspai   = 少数派 (sspai.com/feed)
 *
 * Usage:
 *   const CNTechNews = require('./CNTechNews');
 *   const news = new CNTechNews();
 *   const result = await news.fetchAll();
 *   const result = await news.fetchAll(10, ['36kr', 'ithome']);
 */

const https = require('https');
const http = require('http');

const TIMEOUT = 15000;
const UA = 'AutoClaw-CNTechNews/1.0';
const DEFAULT_COUNT = 20;

const SOURCE_NAMES = {
  '36kr': '36\u6c2a',
  tmt: '\u949b\u5a92\u4f53',
  ithome: 'IT\u4e4b\u5bb6',
  sspai: '\u5c11\u6570\u6d3e',
};

const RSS_FEEDS = {
  '36kr': 'https://36kr.com/feed',
  tmt: 'https://www.tmtpost.com/rss.xml',
  ithome: 'https://www.ithome.com/rss/',
  sspai: 'https://sspai.com/feed',
};

// ── HTML Stripping ────────────────────────────────────────

function decodeEntities(s) {
  return s.replace(/&#x([0-9a-f]+);/gi, (_, hex) => String.fromCodePoint(parseInt(hex, 16)))
    .replace(/&#(\d+);/g, (_, dec) => String.fromCodePoint(Number(dec)))
    .replace(/&quot;/g, '"').replace(/&apos;/g, "'").replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>').replace(/&amp;/g, '&');
}

function stripHtml(s) {
  if (!s) return '';
  return decodeEntities(s.replace(/<[^>]*>/g, '').replace(/\s+/g, ' ')).trim();
}

function clean(s) {
  if (!s) return null;
  s = stripHtml(s);
  return s.length > 500 ? s.substring(0, 500) : s;
}

// ── HTTP ──────────────────────────────────────────────────

function httpGet(url, redirects = 0) {
  const mod = url.startsWith('https') ? https : http;
  return new Promise((resolve, reject) => {
    const req = mod.get(url, {
      headers: {
        'User-Agent': UA,
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
        'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8',
      },
      timeout: TIMEOUT,
    }, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        res.resume();
        if (redirects >= 5) return reject(new Error('Too many redirects'));
        const loc = res.headers.location;
        const next = loc.startsWith('http') ? loc : new URL(loc, url).href;
        httpGet(next, redirects + 1).then(resolve, reject);
        return;
      }
      if (res.statusCode >= 400) {
        res.resume();
        return reject(new Error(`HTTP ${res.statusCode}`));
      }
      res.setEncoding('utf8');
      let data = '';
      res.on('data', (chunk) => { data += chunk; });
      res.on('end', () => resolve(data));
    });
    req.on('error', reject);
    req.on('timeout', () => { req.destroy(); reject(new Error('Request timed out')); });
  });
}

// ── RSS Parser ────────────────────────────────────────────

function extractTag(block, tag) {
  const cdataRe = new RegExp(`<(?:[\\w:]*:)?${tag}[^>]*>\\s*<!\\[CDATA\\[([\\s\\S]*?)\\]\\]>\\s*<\\/`, 'i');
  const cdataMatch = cdataRe.exec(block);
  if (cdataMatch) return cdataMatch[1].trim();

  const re = new RegExp(`<(?:[\\w:]*:)?${tag}[^>]*>([\\s\\S]*?)<\\/(?:[\\w:]*:)?${tag}>`, 'i');
  const m = re.exec(block);
  return m ? m[1].trim() : null;
}

function extractAttr(block, tag, attr) {
  const re = new RegExp(`<${tag}[^>]+${attr}="([^"]+)"`, 'i');
  const m = re.exec(block);
  return m ? m[1] : null;
}

function parseRss(xml, sourceName, n) {
  const articles = [];
  const itemRegex = /<(?:item|entry)[\s>]([\s\S]*?)<\/(?:item|entry)>/gi;
  let match;

  while ((match = itemRegex.exec(xml)) !== null && articles.length < n) {
    const block = match[1];
    const title = extractTag(block, 'title');
    if (!title) continue;

    let link = extractAttr(block, 'link', 'href') || extractTag(block, 'link');
    const description = clean(
      extractTag(block, 'description') || extractTag(block, 'summary') || extractTag(block, 'encoded')
    );
    const pubDate = extractTag(block, 'pubDate') || extractTag(block, 'published') || extractTag(block, 'updated');
    const author = extractTag(block, 'creator') || extractTag(block, 'author');

    const a = { source: sourceName, title: stripHtml(title).trim(), url: (link || '').trim() };
    if (description) a.description = description.length > 300 ? description.substring(0, 300) : description;
    if (pubDate) a.published = pubDate.trim();
    if (author) a.author = stripHtml(author).trim();
    if (a.title) articles.push(a);
  }
  return articles;
}

// ── Formatters ────────────────────────────────────────────

function fmtBrief(allResults) {
  let sb = '';
  let idx = 1;
  for (const [name, articles] of Object.entries(allResults)) {
    for (const a of articles) {
      if (a.error || !a.title) continue;
      sb += `${idx}. [${name}] ${a.title}`;
      if (a.description) {
        sb += ' -- ' + (a.description.length > 150 ? a.description.substring(0, 150) : a.description);
      }
      sb += '\n';
      idx++;
    }
  }
  const now = new Date().toLocaleString('zh-CN', { year: 'numeric', month: '2-digit', day: '2-digit', hour: '2-digit', minute: '2-digit', hour12: false });
  return `\u4e2d\u56fd\u79d1\u6280\u65b0\u95fb (${now}) - ${idx - 1} \u7bc7\u6587\u7ae0\n${sb}`;
}

function fmtDetailed(allResults) {
  let sb = '';
  let total = 0;
  for (const [name, articles] of Object.entries(allResults)) {
    const valid = articles.filter(a => !a.error && a.title);
    const errors = articles.filter(a => a.error);
    sb += `\n## ${name} (${valid.length} \u7bc7)\n`;
    for (const e of errors) sb += `  [\u9519\u8bef: ${e.error}]\n`;
    let i = 1;
    for (const a of valid) {
      sb += `  ${i}. **${a.title}**\n`;
      if (a.url) sb += `     ${a.url}\n`;
      if (a.description) {
        sb += `     ${a.description.length > 250 ? a.description.substring(0, 250) : a.description}\n`;
      }
      if (a.published) sb += `     \u53d1\u5e03: ${a.published}\n`;
      i++; total++;
    }
  }
  const now = new Date().toLocaleString('zh-CN', { year: 'numeric', month: '2-digit', day: '2-digit', hour: '2-digit', minute: '2-digit', hour12: false });
  return `# \u4e2d\u56fd\u79d1\u6280\u65b0\u95fb - ${now}\n**${total} \u7bc7\u6587\u7ae0\uff0c\u6765\u81ea ${Object.keys(allResults).length} \u4e2a\u6e90**\n${sb}`;
}

// ── Public API ────────────────────────────────────────────

class CNTechNews {
  /** Fetch all sources, 20 articles each. */
  async fetchAll(count = DEFAULT_COUNT, sources = null) {
    const results = {};
    const targets = sources || Object.keys(SOURCE_NAMES);

    for (const key of targets) {
      const name = SOURCE_NAMES[key];
      const feed = RSS_FEEDS[key];
      if (!name || !feed) continue;
      try {
        const xml = await httpGet(feed);
        results[name] = parseRss(xml, name, count);
      } catch (e) {
        results[name] = [{ source: name, error: e.message }];
      }
    }

    return {
      results,
      briefSummary: fmtBrief(results),
      detailedSummary: fmtDetailed(results),
      toJson() {
        const o = {};
        for (const [k, v] of Object.entries(this.results)) o[k] = v;
        return o;
      },
    };
  }

  /** Fetch a single source by code (36kr, tmt, ithome, sspai). */
  async fetchSource(code, count = DEFAULT_COUNT) {
    return this.fetchAll(count, [code]);
  }
}

module.exports = CNTechNews;
